src/jvmMain/kotlin/com/sdercolin/vlabeler/video/VideoMain.kt

```diff
@@ -0,0 +1,39 @@
+package com.sdercolin.vlabeler.video
+
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.DisposableEffect
+import com.sdercolin.vlabeler.audio.PlayerState
+import com.sdercolin.vlabeler.env.Log
+import com.sdercolin.vlabeler.util.toMillisecond
+import com.sdercolin.vlabeler.util.or
+
+@Composable
+fun VideoMain(
+    videoState: VideoState,
+    playerState: PlayerState
+) {
+    require(videoState.videoPath != "") { "video path not located" }
+
+    val cause = when (listOf(playerState.isPlaying, videoState.afterPopup)) {
+        listOf(true, false) -> SyncCause.OpenDuringPlay
+        listOf(true, true) -> SyncCause.PlayerStartPlay
+        listOf(false, false) -> SyncCause.RecoverFromLastExit
+        listOf(false, true) -> SyncCause.CloseDuringPlay
+        else -> SyncCause.Nothing
+    }
+    videoState.syncOp = videoState.sync(cause, playerState.startFrameToken?.let {
+        toMillisecond(it.toFloat(), videoState.currentSampleRate).toLong()
+    }.or(videoState.currentTime))
+
+    when (videoState.embeddedMode) {
+        true -> embeddedMode(videoState)
+        false -> newWindowMode(videoState)
+    }
+    videoState.afterPopup = true
```

Comment:
We cannot write side effects directly in composable functions.

I guess you want to do:
```kotlin
DisposableEffect(Unit) {
    videoState.afterPopup = true
    onDispose {
        videoState.afterPopup = false
        videoState.miniVideo.toggleCallback(false)
    }
}
```

Reply:
I wanted to say `LaunchEffect(videoState.mode)` (every time the mode changes, re-execute the code) but didn't realize I could do so. This weird boolean flag can be wrapped inside `LaunchEffect(key)`.